lib/compat/wordpress-6.1/block-template-utils.php

```diff
@@ -27,3 +27,267 @@ function gutenberg_get_default_block_template_types( $default_template_types ) {
+		}
+	}
+	/**
+	 * Filters the array of queried block templates array after they've been fetched.
+	 *
+	 * @since 10.8
+	 *
+	 * @param Gutenberg_Block_Template[] $query_result Array of found block templates.
+	 * @param array $query {
+	 *     Optional. Arguments to retrieve templates.
+	 *
+	 *     @type array  $slug__in List of slugs to include.
+	 *     @type int    $wp_id Post ID of customized template.
+	 * }
+	 * @param array $template_type wp_template or wp_template_part.
+	 */
+	return apply_filters( 'get_block_templates', $query_result, $query, $template_type );
+}
+
+/**
+ * Retrieves a single unified template object using its id.
+ *
+ * @param string $id Template unique identifier (example: theme_slug//template_slug).
+ * @param array  $template_type wp_template or wp_template_part.
+ *
+ * @return Gutenberg_Block_Template|null Template.
+ */
+function gutenberg_get_block_template( $id, $template_type = 'wp_template' ) {
+	/**
+	 * Filters the block template object before the query takes place.
+	 *
+	 * Return a non-null value to bypass the WordPress queries.
+	 *
+	 * @since 10.8
+	 *
+	 * @param Gutenberg_Block_Template|null $block_template Return block template object to short-circuit the default query,
+	 *                                               or null to allow WP to run it's normal queries.
+	 * @param string $id Template unique identifier (example: theme_slug//template_slug).
+	 * @param array  $template_type wp_template or wp_template_part.
+	 */
+	$block_template = apply_filters( 'pre_get_block_template', null, $id, $template_type );
+	if ( ! is_null( $block_template ) ) {
+		return $block_template;
+	}
+
+	$parts = explode( '//', $id, 2 );
+	if ( count( $parts ) < 2 ) {
+		return null;
+	}
+	list( $theme, $slug ) = $parts;
+	$wp_query_args        = array(
+		'post_name__in'  => array( $slug ),
+		'post_type'      => $template_type,
+		'post_status'    => array( 'auto-draft', 'draft', 'publish', 'trash' ),
+		'posts_per_page' => 1,
+		'no_found_rows'  => true,
+		'tax_query'      => array(
+			array(
+				'taxonomy' => 'wp_theme',
+				'field'    => 'name',
+				'terms'    => $theme,
+			),
+		),
+	);
+	$template_query       = new WP_Query( $wp_query_args );
+	$posts                = $template_query->posts;
+
+	if ( count( $posts ) > 0 ) {
+		$template = gutenberg_build_block_template_result_from_post( $posts[0] );
+
+		if ( ! is_wp_error( $template ) ) {
+			return $template;
+		}
+	}
+
+	$block_template = get_block_file_template( $id, $template_type );
+
+	/**
+	 * Filters the queried block template object after it's been fetched.
+	 *
+	 * @since 10.8
+	 *
+	 * @param Gutenberg_Block_Template|null $block_template The found block template, or null if there isn't one.
+	 * @param string $id Template unique identifier (example: theme_slug//template_slug).
+	 * @param array  $template_type wp_template or wp_template_part.
+	 */
+	return apply_filters( 'get_block_template', $block_template, $id, $template_type );
```

Comment:
Unrelated to this PR, but this filter will never apply if we have template object results from DB. See `is_wp_error` condition above.

Reply:
I noticed that as well, but it might be better to handle this separately. WDYT?

Reply:
Also it might be intentional, I'm not really sure and would need checking.

Reply:
Sure. Let's fix this in a follow-up.